### PR TITLE
[#977] `Devshards`: limit inference count

### LINKED ISSUE #977: [P0] `devshards`: Limit amount of inferences
As described in https://github.com/gonka-ai/gonka/issues/914#issuecomment-4090483233, we want to limit the amount of inferences that can be done in a `devshard`, as a way of limiting the amount of damage a hijacked `devshard` can cause on the network.

For now, let's say each `devshard` can run up to 2k inferences.

Upon settlement, the protocol should verify "Missed inferences + Invalidated inferences" does not exceed 2k, for the whole group.

### PR DESCRIPTION
**Pull request by dcastro**
This PR adds a limit of 2k inferences per subnet.

As such, the "Invalid" / "Missed" stats cannot have a total above 2k either. This creates a limit on how much damage a hijacked subnet can cause on the network.

Fixes #977

### CODE REVIEW

**Comment by akup** [comment]
This PR also introduces changes that need consensus sync. While now subnets are still experimental and users will not be affected, we need to think in advance how to make upgrades in next versions when subnets will be opened for wide usage.
There is a proposal on versioning:
https://github.com/gonka-ai/gonka/tree/gm/subnet/proposals/versioned

My concern is:

> how it will handle protocol changes in running subnet?
I mean that all subnet nodes should change the version simultaneously or consensus will be broken. When we used it in cosomosSDK it was binded to block height, and we was sure that on next height everybody will process next block with same version.
But now different subnet nodes can have messages by gossiping and directly from user, and each node can be unsynched with a chain and have different block height, and can download and restart a new version.
Another case is that Escrow finalization protocol changes, and subnet shouldn’t submit finalization in new format before network update, and shouldn’t submit old format after inference-chain upgrade
I didn’t see it is covered by the proposal

@gmorgachev I think there should be new discussion thread on versioning of subnets proposal for public discussion

**Review by akup — CHANGES_REQUESTED** [review]
# PR #1005 — Summarized findings

## Summary

PR #1005 adds a **per-subnet inference cap** (~2000) on L2 (`SessionConfig.MaxInferencesPerSubnet` + `applyStartInference`) and an L1 **settlement gate** (`MaxSettlementInferenceCount`, `verifyHostStatsIntegrity`). Reviewers treat this as **high blast radius**: **wire/proto drift** can zero the new config at runtime, **L1 hardcoded vs L2-configurable limits** can let sessions pass on L2 but **fail settlement** (escrow / economic deadlock), and **arithmetic / typing** around host stats can **overflow** or behave differently across architectures. Tests and docs need small hardening.

## Must fix:

- **Protobuf (and regen) for `MaxInferencesPerSubnet`** — If the field exists on `SessionConfig` in Go but not in `.proto`, serialization/gossip will **drop** it and it will deserialize as **0**, breaking the inference limit check and stalling subnets (`subnet/types/domain.go`; update `subnet/proto` / generated code accordingly).
- **Single source of truth for the cap** — **L1** uses a hardcoded settlement max while **L2** allows a configurable `MaxInferencesPerSubnet`; if L2 can exceed L1, hosts can execute work that **L1 settlement rejects**, risking **stuck escrow** and inconsistent policy. Align via a **shared constant**, **chain params / escrow metadata**, or **strict validation at subnet/session creation** so L2 never exceeds what L1 will accept.

> Seams we need to make MaxInferencesPerSubnet parameter managed via governance proposal

- **Overflow-safe `sumCounts`** — Summing `Validated` + `Invalid` + `Missed` with plain `+` on `uint64` can **wrap** and defeat checks vs `InferenceCount`. Use **`math/bits.Add64`** (or equivalent) and handle carry (`subnet_settlement.go` ~162–165).
- **Deterministic limit comparison** — Avoid `int(len(...))` vs `uint32` cap; compare with **consistent unsigned widths** (e.g. `uint32(len(sm.state.Inferences)) >= config.MaxInferencesPerSubnet`) so **32-bit vs 64-bit** hosts cannot disagree (`subnet/state/machine.go` ~554).
- **Upgrade / legacy sessions** — Old persisted state without the field will read as **0** and can **block all new inferences** after upgrade. Provide a **non-zero default**, **migration**, or **versioned decode** so existing sessions keep working.

## Should fix:

- **Cap untrusted config** — Bound **`MaxInferencesPerSubnet`** with a **network maximum** (constant or params) at config acceptance time so sequencers cannot set **huge** values and **DoS** hosts via unbounded `Inferences` growth (`subnet/state/machine.go`, `domain.go`).
- **Rollout gating** — If the new **reject path** in `applyStartInference` can run on a mix of binary versions, risk **state root divergence** until everyone upgrades; use a **protocol/version gate** or coordinated rollout (`subnet/state/machine.go` ~554).

> This point in my opinion can be skipped before subnet adoption and should be discussed in `devshards` upgrade and versioning thread

- **Nil-safe settlement iteration** — Guard **`hs == nil`** (or invalid entries) before reading `SlotId` in `verifyHostStatsIntegrity` / host stat loops to avoid **validator panic DoS** (`subnet_settlement.go` ~147–150).
- **Tests: global `sdk` config** — Move **`sdk.GetConfig().SetBech32PrefixForAccount`** (and similar) to **`TestMain`** / scoped setup to avoid **parallel test races** (`msg_server_settle_subnet_escrow_test.go` ~410).
- **Semantics review** — Revisit strict per-host / global checks for **griefing** (forced settlement failure after work); confirm **`InferenceCount` vs sum of categories** should use `>` vs `!=` per product rules (`subnet_settlement.go`).

## Nits:

- **Docstring** — `verifyHostStatsIntegrity` comment should match behavior (it returns **`error`**, not “total cost”) (`subnet_settlement.go` ~142).
- **Performance** — If `VerifySubnetSettlement` walks **`HostStats` twice**, consider **one pass** if still hot enough to matter (`subnet_settlement.go` ~117–124).
- **Tests** — Prefer **derived boundary values** from constants over magic numbers; trim **heavy iteration** in setup where possible (`subnet_settlement_test.go`, `machine_test.go`).

File: inference-chain/x/inference/keeper/subnet_settlement.go [hunk 1]
```diff
@@ -136,6 +139,45 @@ func VerifySubnetSettlement(escrow types.SubnetEscrow, msg *types.MsgSettleSubne
 	return nil
 }
 
+// verifyHostStatsIntegrity validates host stats entries and returns total cost.
```
[the remark this answers is not in the document]
Comment (dcastro) [reply inference-chain/x/inference/keeper/subnet_settlement.go]: Done

File: inference-chain/x/inference/keeper/subnet_settlement.go [hunk 2]
```diff
@@ -136,6 +139,45 @@ func VerifySubnetSettlement(escrow types.SubnetEscrow, msg *types.MsgSettleSubne
...
+		// NOTE: It's important that we check that the sum of `Validated/Invalid/Missed` for all hosts
+		// does not exceed [MaxSettlementInferenceCount]
+		// in order to limit the amount of damage a hijacked subnet could cause by reporting fake stats.
+		//
+		// See: https://github.com/gonka-ai/gonka/issues/914#issuecomment-4090483233
+
+		// Per-host consistency: validated + invalid + missed must not exceed inference_count
+		sumCounts := uint64(hs.Validated) + uint64(hs.Invalid) + uint64(hs.Missed)
+		if sumCounts > uint64(hs.InferenceCount) {
+			return fmt.Errorf("host_stats slot_id %d: validated+invalid+missed (%d) exceeds inference_count %d", hs.SlotId, sumCounts, hs.InferenceCount)
```
[the remark this answers is not in the document]
Comment (dcastro) [reply inference-chain/x/inference/keeper/subnet_settlement.go]: Done

File: subnet/state/machine_test.go [hunk 3]
```diff
@@ -102,6 +102,42 @@ func TestApplyDiff_StartInference(t *testing.T) {
...
+			InferenceId: nonce,
+			PromptHash:  []byte("prompt"),
+			Model:       "llama",
+			InputLength: 0,
+			MaxTokens:   0,
+			StartedAt:   1000,
+		})})
+		_, err := sm.ApplyDiff(diff)
+		require.NoError(t, err)
+	}
```
[the remark this answers is not in the document]
Comment (dcastro) [reply subnet/state/machine_test.go:131]: Done

**Comment by akup** [comment]
> This PR also introduces changes that need consensus sync. While now subnets are still experimental and users will not be affected, we need to think in advance how to make upgrades in next versions when subnets will be opened for wide usage. There is a proposal on versioning: https://github.com/gonka-ai/gonka/tree/gm/subnet/proposals/versioned
> 
> My concern is:
> 
> > how it will handle protocol changes in running subnet?
> > I mean that all subnet nodes should change the version simultaneously or consensus will be broken. When we used it in cosomosSDK it was binded to block height, and we was sure that on next height everybody will process next block with same version.
> > But now different subnet nodes can have messages by gossiping and directly from user, and each node can be unsynched with a chain and have different block height, and can download and restart a new version.
> > Another case is that Escrow finalization protocol changes, and subnet shouldn’t submit finalization in new format before network update, and shouldn’t submit old format after inference-chain upgrade
> > I didn’t see it is covered by the proposal
> 
> @gmorgachev I think there should be new discussion thread on versioning of subnets proposal for public discussion

For transparency. Running subnet will not be updated, when escrow is started version of subnet is fixed. Hosts will be able to run simultaneously different subnet versions.
Mainnet (`inference-chain`) should support multiple versions of devshards settlement logics, and at `MsgSettleSubnetEscrow` it should be transfered `devshard` version.

**Review by akup — APPROVED** [review]
I don't like that `SessionConfig` has hardcoded `MaxInferencesPerSubnet` parameter.
But other params are also hardcoded.
`SessionConfig` params definitely should be settable from mainnet on start, and should be configurable by government. But this issue can be given for community to fix https://github.com/gonka-ai/gonka/issues/1028.

**Comment by a-kuprin** [comment]
Implemented via `max-nonce` at 0.2.13 at chain side and at
https://github.com/gonka-ai/gonka/pull/1258 at devshardd side
